### ISSUE TITLE
Assertion `ClsMembers.contains(Old)' failed with clang -cc1 -analyze -analyzer-checker=core

### ISSUE THREAD

**Issue by mikaelholmen** [llvm/llvm-project#55546]
llvm commit: 573a5b580
Reproduce with:
```shell
clang -cc1 -analyze -analyzer-checker=core -o /dev/null bbi-69854.c
```
[bbi-69854.c.gz](https://github.com/llvm/llvm-project/files/8713977/bbi-69854.c.gz)

Result:
```console
bbi-69854.c:7:23: warning: equality comparison result unused [-Wunused-comparison]
    (a | (1 + c)) / b == d;
    ~~~~~~~~~~~~~~~~~~^~~~
bbi-69854.c:8:7: warning: inequality comparison result unused [-Wunused-comparison]
    c != a;
    ~~^~~~
bbi-69854.c:8:7: note: use '|=' to turn this inequality comparison into an or-assignment
    c != a;
      ^~
      |=
bbi-69854.c:10:11: warning: if statement has empty body [-Wempty-body]
    if (d);
          ^
bbi-69854.c:10:11: note: put the semicolon on a separate line to silence this warning
clang: ../../clang/lib/StaticAnalyzer/Core/RangeConstraintManager.cpp:2509: clang::ento::ProgramStateRef (anonymous namespace)::EquivalenceClass::removeMember(clang::ento::ProgramStateRef, const clang::ento::SymbolRef): Assertion `ClsMembers.contains(Old)' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: build-all/bin/clang -cc1 -analyze -analyzer-checker=core -o /dev/null bbi-69854.c
1.	<eof> parser at end of file
2.	While analyzing stack: 
	#0 Calling f
3.	bbi-69854.c:8:5: Error evaluating statement
4.	bbi-69854.c:8:5: Error evaluating statement
 #0 0x0000000002e0d863 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (build-all/bin/clang+0x2e0d863)
 #1 0x0000000002e0b4de llvm::sys::RunSignalHandlers() (build-all/bin/clang+0x2e0b4de)
 #2 0x0000000002e0dbe6 SignalHandler(int) Signals.cpp:0:0
 #3 0x00007fd5535c6630 __restore_rt sigaction.c:0:0
 #4 0x00007fd550d0d387 raise (/lib64/libc.so.6+0x36387)
 #5 0x00007fd550d0ea78 abort (/lib64/libc.so.6+0x37a78)
 #6 0x00007fd550d061a6 __assert_fail_base (/lib64/libc.so.6+0x2f1a6)
 #7 0x00007fd550d06252 (/lib64/libc.so.6+0x2f252)
 #8 0x0000000004573602 (anonymous namespace)::EquivalenceClass::simplify(clang::ento::SValBuilder&, clang::ento::RangeSet::Factory&, llvm::IntrusiveRefCntPtr<clang::ento::ProgramState const>, (anonymous namespace)::EquivalenceClass) RangeConstraintManager.cpp:0:0
 #9 0x00000000045723f2 (anonymous namespace)::ConstraintAssignorBase<(anonymous namespace)::ConstraintAssignor>::assignSymExprImpl(clang::ento::SymExpr const*, clang::ento::RangeSet) RangeConstraintManager.cpp:0:0
#10 0x000000000456daa7 (anonymous namespace)::RangeConstraintManager::setRange(llvm::IntrusiveRefCntPtr<clang::ento::ProgramState const>, clang::ento::SymExpr const*, clang::ento::RangeSet) RangeConstraintManager.cpp:0:0
#11 0x0000000004563b8c (anonymous namespace)::RangeConstraintManager::assumeSymEQ(llvm::IntrusiveRefCntPtr<clang::ento::ProgramState const>, clang::ento::SymExpr const*, llvm::APSInt const&, llvm::APSInt const&) RangeConstraintManager.cpp:0:0
#12 0x0000000004577e06 clang::ento::RangedConstraintManager::assumeSymRel(llvm::IntrusiveRefCntPtr<clang::ento::ProgramState const>, clang::ento::SymExpr const*, clang::BinaryOperatorKind, llvm::APSInt const&) (build-all/bin/clang+0x4577e06)
#13 0x0000000004577124 clang::ento::RangedConstraintManager::assumeSym(llvm::IntrusiveRefCntPtr<clang::ento::ProgramState const>, clang::ento::SymExpr const*, bool) (build-all/bin/clang+0x4577124)
#14 0x0000000004594b2d clang::ento::SimpleConstraintManager::assumeAux(llvm::IntrusiveRefCntPtr<clang::ento::ProgramState const>, clang::ento::NonLoc, bool) (build-all/bin/clang+0x4594b2d)
#15 0x0000000004594898 clang::ento::SimpleConstraintManager::assume(llvm::IntrusiveRefCntPtr<clang::ento::ProgramState const>, clang::ento::NonLoc, bool) (build-all/bin/clang+0x4594898)
#16 0x00000000045947ab clang::ento::SimpleConstraintManager::assumeInternal(llvm::IntrusiveRefCntPtr<clang::ento::ProgramState const>, clang::ento::DefinedSVal, bool) (build-all/bin/clang+0x45947ab)
#17 0x00000000044e10b5 clang::ento::ConstraintManager::assumeDual(llvm::IntrusiveRefCntPtr<clang::ento::ProgramState const>, clang::ento::DefinedSVal) (build-all/bin/clang+0x44e10b5)
#18 0x0000000004501bed clang::ento::ExprEngine::evalEagerlyAssumeBinOpBifurcation(clang::ento::ExplodedNodeSet&, clang::ento::ExplodedNodeSet&, clang::Expr const*) (build-all/bin/clang+0x4501bed)
#19 0x00000000044ff118 clang::ento::ExprEngine::Visit(clang::Stmt const*, clang::ento::ExplodedNode*, clang::ento::ExplodedNodeSet&) (build-all/bin/clang+0x44ff118)
#20 0x00000000044fa82e clang::ento::ExprEngine::ProcessStmt(clang::Stmt const*, clang::ento::ExplodedNode*) (build-all/bin/clang+0x44fa82e)
#21 0x00000000044fa559 clang::ento::ExprEngine::processCFGElement(clang::CFGElement, clang::ento::ExplodedNode*, unsigned int, clang::ento::NodeBuilderContext*) (build-all/bin/clang+0x44fa559)
#22 0x00000000044e2980 clang::ento::CoreEngine::HandlePostStmt(clang::CFGBlock const*, unsigned int, clang::ento::ExplodedNode*) (build-all/bin/clang+0x44e2980)
#23 0x00000000044e1a1a clang::ento::CoreEngine::ExecuteWorkList(clang::LocationContext const*, unsigned int, llvm::IntrusiveRefCntPtr<clang::ento::ProgramState const>) (build-all/bin/clang+0x44e1a1a)
#24 0x000000000416174c (anonymous namespace)::AnalysisConsumer::HandleCode(clang::Decl*, unsigned int, clang::ento::ExprEngine::InliningModes, llvm::DenseSet<clang::Decl const*, llvm::DenseMapInfo<clang::Decl const*, void>>*) AnalysisConsumer.cpp:0:0
#25 0x0000000004149baa (anonymous namespace)::AnalysisConsumer::HandleTranslationUnit(clang::ASTContext&) AnalysisConsumer.cpp:0:0
#26 0x00000000045db223 clang::ParseAST(clang::Sema&, bool, bool) (build-all/bin/clang+0x45db223)
#27 0x0000000003809826 clang::FrontendAction::Execute() (build-all/bin/clang+0x3809826)
#28 0x000000000377dc54 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (build-all/bin/clang+0x377dc54)
#29 0x00000000038bd0e2 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (build-all/bin/clang+0x38bd0e2)
#30 0x0000000000a1323c cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (build-all/bin/clang+0xa1323c)
#31 0x0000000000a10fcf ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&) driver.cpp:0:0
#32 0x0000000000a10b53 main (build-all/bin/clang+0xa10b53)
#33 0x00007fd550cf9555 __libc_start_main (/lib64/libc.so.6+0x22555)
#34 0x0000000000a0dc0c _start (build-all/bin/clang+0xa0dc0c)
Abort
```
Starts happening with 20f8733d4b:
```text
[Analyzer][solver] Simplification: Do a fixpoint iteration before the eq class merge

This reverts commit f02c5f3478318075d1a469203900e452ba651421 and
addresses the issue mentioned in D114619 differently.
```

**Comment by mikaelholmen** [comment]
cc @martong

**Comment by steakhal** [comment]
Confirmed. Still crashes.

**Comment by martong** [comment]
Here is a smaller and cleaner reproducer:
```
extern void abort() __attribute__((__noreturn__));
#define assert(expr) ((expr) ? (void)(0) : abort())

int a, b, d;
long c; // int c; does not crash
void test() {
  assert(a + 1 + c + b != d);
  assert(c == a);
  assert(d == 0); // d = 0; does not crash
  c = 0;

  if ((a + 1 + c + b == c) != a)
    ;
}
```
